### PR TITLE
Handle a NOOP state transition request

### PR DESCRIPTION
**Pull request by brianjmurrell**
Fixes HYD-3468

When trying to stop a filesystem that has a worker configured but not
started one can get an error such as:

```
[2014-08-19 18:35:56,694: ERROR/django.request.tastypie] Internal Server Error: /api/filesystem/1/
Traceback (most recent call last):
 File "/usr/lib/python2.6/site-packages/tastypie/resources.py", line 192, in wrapper
   response = callback(request, *args, **kwargs)
 File "/usr/lib/python2.6/site-packages/tastypie/resources.py", line 406, in dispatch_detail
   return self.dispatch('detail', request, **kwargs)
 File "/usr/lib/python2.6/site-packages/tastypie/resources.py", line 427, in dispatch
   response = method(request, **kwargs)
 File "/usr/lib/python2.6/site-packages/tastypie/resources.py", line 1131, in put_detail
   updated_bundle = self.obj_update(bundle, request=request, **self.remove_api_resource_names(kwargs))
 File "/usr/share/chroma-manager/chroma_api/utils.py", line 319, in obj_update
 File "/usr/share/chroma-manager/chroma_api/utils.py", line 267, in obj_update
 File "/usr/share/chroma-manager/chroma_core/services/job_scheduler/job_scheduler_client.py", line 203, in get_transition_consequences
 File "/usr/share/chroma-manager/chroma_core/services/job_scheduler/command_plan.py", line 176, in get_transition_consequences
 File "/usr/share/chroma-manager/chroma_core/services/job_scheduler/command_plan.py", line 377, in _emit_transition_deps
 File "/usr/share/chroma-manager/chroma_core/services/job_scheduler/command_plan.py", line 456, in _collect_dependencies
 File "/usr/share/chroma-manager/chroma_core/services/job_scheduler/command_plan.py", line 369, in _emit_transition_deps
 File "/usr/share/chroma-manager/chroma_core/models/jobs.py", line 259, in get_route SchedulingError: unmounted->unmounted not legal state transition
for <class 'chroma_core.models.client_mount.LustreClientMount'>
```

This appears to be because our state engine does not handle NOOP state
transition requests.

Signed-off-by: Brian J. Murrell <brian.murrell@intel.com>

---
This change is [Reviewable](https://reviewable.io/reviews/intel-hpdd/intel-manager-for-lustre/294)

### CODE REVIEW

File: chroma-manager/chroma_core/services/job_scheduler/command_plan.py [hunk 1]
```diff
@@ -346,6 +346,12 @@ def _emit_transition_deps(self, transition, transition_stack = {}):
         transition_stack[transition.stateful_object] = transition.new_state
         log.debug("Updating transition_stack[%s/%s] = %s" % (transition.stateful_object.__class__, transition.stateful_object.id, transition.new_state))
 
+        # do nothing for a NOOP transition
+        if transition.old_state == transition.new_state:
```
Comment (jgrund) [inline chroma-manager/chroma_core/services/job_scheduler/command_plan.py]: This seems like a gap we would have seen many times. Are you sure there's not some other edge case being missed?

Comment (brianjmurrell) [reply chroma-manager/chroma_core/services/job_scheduler/command_plan.py]: That is **exactly** what I was thinking and I am far from an expert in this area of the code, so the more eyes the better.

But this simple fix does make [test_mount_and_unmount](https://github.com/intel-hpdd/intel-manager-for-lustre/blob/92f459306fdf0bb26528e8a55a236ab0e35531e2/chroma-manager/tests/integration/shared_storage_configuration/test_client_mount_management.py#L22) pass where it was previously failing with [HYD-3468](https://jira.hpdd.intel.com/browse/HYD-3468).

I've started up 5x SSI testing with this patch underneath #250 so we'll see how it fares.

Comment (brianjmurrell) [reply chroma-manager/chroma_core/services/job_scheduler/command_plan.py]: It had one actual honest pass and the other 4 failures were all for other known intermittent issues.

File: chroma-manager/chroma_core/services/job_scheduler/command_plan.py [hunk 2]
```diff
@@ -346,6 +346,12 @@ def _emit_transition_deps(self, transition, transition_stack = {}):
...
+            log.debug("NOOP transition %s -> %s" % (transition.old_state, transition.new_state))
+            return None
+
+        route = transition.stateful_object.get_route(transition.old_state, transition.new_state)
```
Comment (tanabarr) [inline chroma-manager/chroma_core/services/job_scheduler/command_plan.py]: why is this line added?

Comment (brianjmurrell) [reply chroma-manager/chroma_core/services/job_scheduler/command_plan.py]: Mistakenly I would say.  I will push a fix for that.  Given that it just duplicated the line below it, remving it shouldn't have any effect, but it should be removed all the same.

**Comment by tanabarr** [comment]
the "handle-noop-transition" change introduces a "pass" for same state transitions into the state machine. this was not previously introduced because the idea is that the state machine should not be issuing same state transitions, and that if it does , then something is broken. this is just an observation, I don't have a strong opinion against doing this, it's just something to be aware of

**Comment by jgrund** [comment]
> the "handle-noop-transition" change introduces a "pass" for same state transitions into the state machine. this was not previously introduced because the idea is that the state machine should not be issuing same state transitions, and that if it does , then something is broken. this is just an observation, I don't have a strong opinion against doing this, it's just something to be aware of

Understood. Given this patch there is something broken consistently. My thoughts are we should land this and then investigate a more "proper" fix post 4.0
